libs/ml/include/ml/graph.hpp

```diff
@@ -218,28 +223,69 @@ void Graph<ArrayType>::ApplyRegularisation()
  */
 template <typename ArrayType>
 template <class OperationType, typename... Params>
-std::string Graph<ArrayType>::AddNode(std::string const &             node_name,
-                                      std::vector<std::string> const &inputs, Params... params)
+meta::IfIsShareable<ArrayType, OperationType, std::string> Graph<ArrayType>::AddNode(
+    std::string const &node_name, std::vector<std::string> const &inputs, Params... params)
 {
   // guarantee unique op name
-  std::string name = UpdateVariableName<OperationType>(node_name);
+  auto        naming_result = UpdateVariableName<OperationType>(node_name);
+  bool        is_duplicate  = naming_result.first;
+  std::string updated_name  = naming_result.second;
 
-  // Instantiate the node
-  auto op      = std::make_shared<Node<ArrayType, OperationType>>(name, params...);
-  nodes_[name] = op;
+  std::shared_ptr<fetch::ml::Node<ArrayType, OperationType>> op;
+  if (!is_duplicate)
+  {
+    // Instantiate the node based on params
+    op = std::make_shared<Node<ArrayType, OperationType>>(updated_name, params...);
+  }
+  else
+  {  // if shared weight is specified by duplicate naming
+    // Instantiate the node based on pointer to shared target node
+    NodePtrType target_node = GetNode(node_name);
+
+    op = std::make_shared<Node<ArrayType, OperationType>>(updated_name, target_node, params...);
+  }
+  nodes_[updated_name] = op;
 
   // assign inputs and outputs
   for (auto const &i : inputs)
   {
-    nodes_[name]->AddInput(nodes_[i]);
-    nodes_[i]->AddOutput(nodes_[name]);
+    nodes_[updated_name]->AddInput(nodes_[i]);
+    nodes_[i]->AddOutput(nodes_[updated_name]);
   }
 
   // add to map of trainable ops if necessary
-  AddTrainable(name, op);
+  AddTrainable(updated_name, op);
 
   // return unique node name (may not be identical to node_name)
-  return name;
+  return updated_name;
+}
+
+template <typename ArrayType>
+template <class OperationType, typename... Params>
+meta::IfIsNotShareable<ArrayType, OperationType, std::string> Graph<ArrayType>::AddNode(
+    std::string const &node_name, std::vector<std::string> const &inputs, Params... params)
+{
+  // guarantee unique op name
+  auto naming_result = UpdateVariableName<OperationType>(node_name);
+
+  std::string updated_name = naming_result.second;
+
+  auto op = std::make_shared<Node<ArrayType, OperationType>>(updated_name, params...);
+
+  nodes_[updated_name] = op;
```

Comment:
lines 275 - 285 are identical for both IfIsShareable and IfIsNotShareable;
Would be nice to make a private member function that does this and just call it from both so as not to have repeated code